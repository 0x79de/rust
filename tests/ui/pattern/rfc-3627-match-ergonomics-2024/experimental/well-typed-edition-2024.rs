//@ edition: 2024
//@ revisions: classic2024 structural2024
//@ run-pass
//! Test cases for well-typed patterns in edition 2024. These are in their own file to ensure we
//! pass both HIR typeck and MIR borrowck, as we may skip the latter if grouped with failing tests.
#![allow(incomplete_features, unused_mut)]
#![cfg_attr(classic2024, feature(ref_pat_eat_one_layer_2024))]
#![cfg_attr(structural2024, feature(ref_pat_eat_one_layer_2024_structural))]

pub fn main() {
    if let Some(Some(&x)) = &Some(&Some(0)) {
        let _: u32 = x;
    }
    if let Some(Some(&x)) = &Some(Some(&0)) {
        let _: &u32 = x;
    }
    if let Some(Some(&&x)) = &Some(Some(&0)) {
        let _: u32 = x;
    }
    if let Some(&Some(x)) = &Some(Some(0)) {
        let _: u32 = x;
    }
    if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
        let _: u32 = x;
    }
    if let Some(Some(&x)) = &Some(&Some(0)) {
        let _: u32 = x;
    }
    if let Some(&Some(&x)) = &mut Some(&Some(0)) {
        let _: u32 = x;
    }
    if let Some(&Some(x)) = &mut Some(&Some(0)) {
        let _: &u32 = x;
    }
    if let Some(&Some(&mut ref x)) = Some(&Some(&mut 0)) {
        let _: &u32 = x;
    }
    if let Some(&Some(&x)) = &Some(&mut Some(0)) {
        let _: u32 = x;
    }
    if let Some(&Some(&x)) = &Some(&Some(0)) {
        let _: u32 = x;
    }
    if let Some(&Some(&x)) = &Some(&mut Some(0)) {
        let _: u32 = x;
    }
    if let Some(&Some(Some(&x))) = &Some(Some(&mut Some(0))) {
        let _: u32 = x;
    }
    if let Some(&Some(&x)) = Some(&Some(&mut 0)) {
        let _: u32 = x;
    }
    if let Some(&Some(x)) = &mut Some(Some(0)) {
        let _: u32 = x;
    }

    // Tests for eat-inner rulesets matching on the outer reference if matching on the inner
    // reference causes a mutability mismatch, i.e. `Deref(EatInner, FallbackToOuter)`:
    let [&mut x] = &mut [&0];
    let _: &u32 = x;

    let [&mut ref x] = &mut [&0];
    let _: &&u32 = x;

    let [&mut ref mut x] = &mut [&0];
    let _: &mut &u32 = x;

    let [&mut mut x] = &mut [&0];
    let _: &u32 = x;

    let [&mut &x] = &mut [&0];
    let _: u32 = x;

    let [&mut &ref x] = &mut [&0];
    let _: &u32 = x;

    let [&mut &(mut x)] = &mut [&0];
    let _: u32 = x;
}
