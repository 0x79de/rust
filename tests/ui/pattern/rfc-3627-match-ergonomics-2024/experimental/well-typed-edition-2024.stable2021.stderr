error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:14:22
   |
LL |     if let Some(Some(&x)) = &Some(&Some(0)) {
   |                      ^^     --------------- this expression has type `&Option<&Option<{integer}>>`
   |                      |
   |                      expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL |     if let Some(Some(x)) = &Some(&Some(0)) {
   |                      ~

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:23:23
   |
LL |     if let Some(Some(&&x)) = &Some(Some(&0)) {
   |                       ^^     --------------- this expression has type `&Option<Option<&{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(Some(&&x)) = &Some(Some(&0)) {
LL +     if let Some(Some(&x)) = &Some(Some(&0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:28:17
   |
LL |     if let Some(&Some(x)) = &Some(Some(0)) {
   |                 ^^^^^^^^    -------------- this expression has type `&Option<Option<{integer}>>`
   |                 |
   |                 expected `Option<{integer}>`, found `&_`
   |
   = note:   expected enum `Option<{integer}>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:33:22
   |
LL |     if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
   |                      ^^^^^^     ----------------------- this expression has type `&mut Option<&mut Option<{integer}>>`
   |                      |
   |                      expected integer, found `&mut _`
   |
   = note:           expected type `{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:33:22
   |
LL |     if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
   |                      ^^^^^^
help: consider removing `&mut` from the pattern
   |
LL |     if let Some(Some(x)) = &mut Some(&mut Some(0)) {
   |                      ~

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:38:23
   |
LL |     if let Some(&Some(&x)) = &mut Some(&Some(0)) {
   |                       ^^     ------------------- this expression has type `&mut Option<&Option<{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL |     if let Some(&Some(x)) = &mut Some(&Some(0)) {
   |                       ~

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:50:17
   |
LL |     if let Some(&Some(&x)) = &Some(&mut Some(0)) {
   |                 ^^^^^^^^^    ------------------- this expression has type `&Option<&mut Option<{integer}>>`
   |                 |
   |                 types differ in mutability
   |
   = note: expected mutable reference `&mut Option<{integer}>`
                      found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:55:23
   |
LL |     if let Some(&Some(&x)) = &Some(&Some(0)) {
   |                       ^^     --------------- this expression has type `&Option<&Option<{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL |     if let Some(&Some(x)) = &Some(&Some(0)) {
   |                       ~

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:60:17
   |
LL |     if let Some(&Some(Some(&x))) = &Some(Some(&mut Some(0))) {
   |                 ^^^^^^^^^^^^^^^    ------------------------- this expression has type `&Option<Option<&mut Option<{integer}>>>`
   |                 |
   |                 expected `Option<&mut Option<{integer}>>`, found `&_`
   |
   = note:   expected enum `Option<&mut Option<{integer}>>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:65:23
   |
LL |     if let Some(&Some(&x)) = Some(&Some(&mut 0)) {
   |                       ^^     ------------------- this expression has type `Option<&Option<&mut {integer}>>`
   |                       |
   |                       types differ in mutability
   |
   = note: expected mutable reference `&mut {integer}`
                      found reference `&_`
help: consider removing `&` from the pattern
   |
LL |     if let Some(&Some(x)) = Some(&Some(&mut 0)) {
   |                       ~

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:70:17
   |
LL |     if let Some(&Some(x)) = &mut Some(Some(0)) {
   |                 ^^^^^^^^    ------------------ this expression has type `&mut Option<Option<{integer}>>`
   |                 |
   |                 expected `Option<{integer}>`, found `&_`
   |
   = note:   expected enum `Option<{integer}>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:78:10
   |
LL |     let [&mut x] = &mut [&0];
   |          ^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:78:10
   |
LL |     let [&mut x] = &mut [&0];
   |          ^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut x] = &mut [&0];
LL +     let [x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:83:10
   |
LL |     let [&mut ref x] = &mut [&0];
   |          ^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:83:10
   |
LL |     let [&mut ref x] = &mut [&0];
   |          ^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut ref x] = &mut [&0];
LL +     let [ref x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:88:10
   |
LL |     let [&mut ref mut x] = &mut [&0];
   |          ^^^^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:88:10
   |
LL |     let [&mut ref mut x] = &mut [&0];
   |          ^^^^^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut ref mut x] = &mut [&0];
LL +     let [ref mut x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:93:10
   |
LL |     let [&mut mut x] = &mut [&0];
   |          ^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:93:10
   |
LL |     let [&mut mut x] = &mut [&0];
   |          ^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut mut x] = &mut [&0];
LL +     let [mut x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:98:10
   |
LL |     let [&mut &x] = &mut [&0];
   |          ^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:103:10
   |
LL |     let [&mut &ref x] = &mut [&0];
   |          ^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:108:10
   |
LL |     let [&mut &(mut x)] = &mut [&0];
   |          ^^^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error: aborting due to 17 previous errors

For more information about this error, try `rustc --explain E0308`.
